phoenix-spark/src/test/java/org/apache/phoenix/spark/datasource/v2/writer/PhoenixTestingDataWriter.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.spark.datasource.v2.writer;
+
+import org.apache.spark.sql.sources.v2.writer.WriterCommitMessage;
+
+import java.sql.SQLException;
+
+public class PhoenixTestingDataWriter extends PhoenixDataWriter {
+
+    private long numBatchesCommitted = 0;
+
+    PhoenixTestingDataWriter(PhoenixDataSourceWriteOptions options) {
+        super(options);
+    }
+
+    // Override to also count the number of times we call this method to test upsert batch commits
```

Comment:
I would not add one-liner comments before the methods if method/variable names are self-justified. Mainly because the comments are likely to get missed from updating when the function or variable changes or gets moved around. Do you think it makes sense?

Could you please also add a brief description before the class?

Reply:
The reason for most of the 1-liner comments is that _why_ the method is being overridden to do what it is doing, is not obvious. The actual overriding part is not being explained. 

As for adding a description for each class, the underlying interface in all cases has in-depth descriptions. For example: 
`PhoenixTestingDataSourceWriter` -> `PhoenixDataSourceWriter` -->  `DataSourceWriter`
and `DataSourceWriter` contains the description. This also avoids redundant comments amongst interfaces and implementing classes. 

Is there anything specific that you would like me to add?